File: CasitaClaude/Services/PeriodService.swift
import Foundation

/// Manages custom billing cycle periods
struct PeriodService {

    /// Calculate the current period range based on the user's billing cycle start day
    static func currentPeriod(startDay: Int) -> (start: Date, end: Date) {
        let calendar = Calendar.current
        let today = Date()
        let currentDay = calendar.component(.day, from: today)

        var startComponents = calendar.dateComponents([.year, .month], from: today)
        let clampedStartDay = min(startDay, 28) // Clamp to 28 to avoid month-length issues

        if currentDay >= clampedStartDay {
            // Current period started this month
            startComponents.day = clampedStartDay
        } else {
            // Current period started last month
            if let lastMonth = calendar.date(byAdding: .month, value: -1, to: today) {
                startComponents = calendar.dateComponents([.year, .month], from: lastMonth)
                startComponents.day = clampedStartDay
            }
        }

        guard let periodStart = calendar.date(from: startComponents),
              let periodEnd = calendar.date(byAdding: .month, value: 1, to: periodStart) else {
            return (today, today)
        }

        // End date is the day before the next period start
        let adjustedEnd = calendar.date(byAdding: .day, value: -1, to: periodEnd) ?? periodEnd

        return (periodStart, adjustedEnd)
    }

    /// Get the previous period range
    static func previousPeriod(startDay: Int) -> (start: Date, end: Date) {
        let calendar = Calendar.current
        let current = currentPeriod(startDay: startDay)

        guard let prevStart = calendar.date(byAdding: .month, value: -1, to: current.start),
              let prevEnd = calendar.date(byAdding: .day, value: -1, to: current.start) else {
            return current
        }

        return (prevStart, prevEnd)
    }

    /// Format a period range for display (e.g., "23 Feb – 22 Mar")
    static func formatPeriod(_ start: Date, _ end: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "es_MX")
        formatter.dateFormat = "d MMM"
        return "\(formatter.string(from: start)) – \(formatter.string(from: end))"
    }

    /// Days remaining in the current period
    static func daysRemaining(startDay: Int) -> Int {
        let calendar = Calendar.current
        let period = currentPeriod(startDay: startDay)
        let today = calendar.startOfDay(for: Date())
        let end = calendar.startOfDay(for: period.end)
        return max(0, calendar.dateComponents([.day], from: today, to: end).day ?? 0)
    }

    /// Days elapsed in the current period
    static func daysElapsed(startDay: Int) -> Int {
        let calendar = Calendar.current
        let period = currentPeriod(startDay: startDay)
        let today = calendar.startOfDay(for: Date())
        let start = calendar.startOfDay(for: period.start)
        return max(0, calendar.dateComponents([.day], from: start, to: today).day ?? 0)
    }

    /// Total days in the current period
    static func totalDays(startDay: Int) -> Int {
        let calendar = Calendar.current
        let period = currentPeriod(startDay: startDay)
        return calendar.dateComponents([.day], from: period.start, to: period.end).day ?? 30
    }
}

File: web/src/services/periodService.js
import { format, addMonths, addDays, differenceInDays, startOfDay } from 'date-fns';
import { es } from 'date-fns/locale';

export function currentPeriod(startDay) {
  const today = new Date();
  const day = today.getDate();
  const clamped = Math.min(startDay, 28);

  let periodStart;
  if (day >= clamped) {
    periodStart = new Date(today.getFullYear(), today.getMonth(), clamped);
  } else {
    const lastMonth = addMonths(today, -1);
    periodStart = new Date(lastMonth.getFullYear(), lastMonth.getMonth(), clamped);
  }

  const nextPeriodStart = addMonths(periodStart, 1);
  const periodEnd = addDays(nextPeriodStart, -1);

  return { start: periodStart, end: periodEnd };
}

export function previousPeriod(startDay) {
  const curr = currentPeriod(startDay);
  const prevStart = addMonths(curr.start, -1);
  const prevEnd = addDays(curr.start, -1);
  return { start: prevStart, end: prevEnd };
}

export function formatPeriod(start, end) {
  const fmt = (d) => format(d, "d MMM", { locale: es });
  return `${fmt(start)} – ${fmt(end)}`;
}

export function daysRemaining(startDay) {
  const period = currentPeriod(startDay);
  const today = startOfDay(new Date());
  const end = startOfDay(period.end);
  return Math.max(0, differenceInDays(end, today));
}

export function daysElapsed(startDay) {
  const period = currentPeriod(startDay);
  const today = startOfDay(new Date());
  const start = startOfDay(period.start);
  return Math.max(0, differenceInDays(today, start));
}

export function totalDays(startDay) {
  const period = currentPeriod(startDay);
  return differenceInDays(period.end, period.start);
}
